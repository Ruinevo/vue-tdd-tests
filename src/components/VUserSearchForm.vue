<template>
  <div class="search-form">
    <input class="search-form__input" type="text" v-model="searchValue" />
    <button class="search-form__submit" @click="$emit('search', searchValue)">Поиск</button>
  </div>
</template>

<script>
export default {
  name: 'VUserSearchForm',
  data: () => ({
    searchValue: ''
  })
}
</script>
