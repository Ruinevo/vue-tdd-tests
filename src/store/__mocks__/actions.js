import userFixture from '@/tests/unit/fixtures/user'

export default {
  /** создаем моковый экшен, который возвращает разрешенный промис со значением фиктивного пользователя */
  searchUser: jest.fn().mockResolvedValue(userFixture)
}
