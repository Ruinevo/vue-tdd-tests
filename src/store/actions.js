import api from '@/api'

export default {
  async searchUser ({ commit }, { username }) {
    try {
      const user = await api.searchUser(username)
      commit('setUser', user)
    } catch (e) {
      console.error(e)
    }
  }
}
