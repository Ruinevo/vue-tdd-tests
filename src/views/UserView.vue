
<template>
  <div>
    <VUserSearchForm @search="search" />
    <VUserProfile :user="user" />
  </div>
</template>

<script>
import VUserProfile from '@/components/VUserProfile'
import VUserSearchForm from '@/components/VUserSearchForm'
import { mapState } from 'vuex';

export default {
  name: 'UserView',
  components: {
    VUserSearchForm,
    VUserProfile
  },
  computed: {
    ...mapState(['user'])
  },
  methods: {
    search (userName) {
      this.$store.dispatch('searchUser', { userName })
    }
  }
};
</script>
