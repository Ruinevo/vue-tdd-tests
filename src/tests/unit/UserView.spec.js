/*eslint-disable */

jest.mock('@/store/actions')

import { shallowMount, createLocalVue } from '@vue/test-utils'
import UserView from '@/views/UserView'
import VUserSearchForm from '@/components/VUserSearchForm'
import VUserProfile from '@/components/VUserProfile'
import Vuex from 'vuex';
import initialState from '@/store/state'
import actions from '@/store/actions'
import userFixture from './fixtures/user'


const localVue = createLocalVue()
localVue.use(Vuex)

describe('UserView', () => {
  let state

  const build = () => {
    const wrapper = shallowMount(UserView, {
      localVue,
      store: new Vuex.Store({ state, actions })
    })
    return {
      wrapper,
      VUserSearchForm: () => wrapper.findComponent(VUserSearchForm),
      VUserProfile: () => wrapper.findComponent(VUserProfile)
    }
  }

  beforeEach(() => {
    jest.resetAllMocks()
    state = { ...initialState }
  })

  it('должен быть отрисован', () => {
    const { wrapper } = build();
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('содержит дочерние компоненты', () => {
    const { VUserSearchForm, VUserProfile } = build();
    expect(VUserSearchForm().exists()).toBe(true)
    expect(VUserProfile().exists()).toBe(true)
  })

  it('передает в дочерний компонент данные из стора', () => {
    state.user = userFixture
    const { VUserProfile } = build()
    expect(VUserProfile().vm.user).toBe(state.user)
  })

  it('при вызове события `search` мы вызываем экшен и передаем туда нашего пользователя', () => {
    const expectedUser = 'Rinat'

    const { VUserSearchForm } = build()
    VUserSearchForm().vm.$emit('search', expectedUser)

    expect(actions.searchUser).toHaveBeenCalled()
    expect(actions.searchUser.mock.calls[0][1]).toEqual({ userName: expectedUser })
  })
})
