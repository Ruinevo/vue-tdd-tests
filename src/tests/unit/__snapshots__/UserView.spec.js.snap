// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserView должен быть отрисован 1`] = `
<div>
  <vusersearchform-stub></vusersearchform-stub>
  <vuserprofile-stub user="[object Object]"></vuserprofile-stub>
</div>
`;
